Workflow improvements

KevinVlaanderen opened the pull request:
## This PR contains
- [ ] Bug fix
- [ ] Feature
- [x] Refactor
- [ ] Documentation
- [ ] Other (describe)

## What is the purpose of this PR?
Various improvements to the state of our repo (specifically automation), ultimately with the goals of:
- speeding up automation (and local builds)
- reducing complexity of pipelines
- removing duplication in pipelines

## Relevant changes
- Migrate to Lerna 5 with NX [5]
- Consistently implement caching (npm and nx) [2]
- Combine workflows with similar triggers into one [3]
- Replace a lot of complex code to determine changed packages with a simpler approach [4]
- Update several of the github actions that are in use [2]
- Enable yml file formatting [1]

## What should be covered while testing?
- Pipeline performance
- Pipeline coverage (is everything properly build/tested?)
- Does partial building work properly? What if you create an MR targetting this branch, with only 1 widget changed?

## Next steps
- Add descriptions/better names to each of the workflows (and jobs/steps, where relevant)
- Reduce complexity of e2e tests; make better use of tools provided by GH actions and nx
- Re-use built artifacts (widgets) in other steps, e.g. e2e tests
- Re-introduce parallelization for e2e tests (it used to run with 3 runners, now only 1)

KevinVlaanderen committed:
chore: use 'lerna bootstrap' instead of 'npm install'

This has two advantages:
- before this, we ran 'npm install' followed by 'lerna bootstrap', which resulted in double work.
Now we just install packages once.
- on ci, 'lerna bootstrap' automatically runs 'npm ci', resulting in faster dependency installation

KevinVlaanderen committed:
chore: use setup-node v3

v3 supports caching, removing the need for an explicit caching step
use all package-locks as input

KevinVlaanderen committed:
chore: remove old compatibility step

KevinVlaanderen committed:
chore: format yml files

KevinVlaanderen committed:
chore: combine lint, build and test in one workflow

KevinVlaanderen committed:
chore: remove 'with:submodules:false', because we do not have submodules

KevinVlaanderen committed:
fix: use matrix.node version for workflow that also runs with node 14

KevinVlaanderen committed:
chore: leverage lerna to identify packages to build/test in Build workflow

KevinVlaanderen committed:
chore: use checkout v3

Some jobs that need other refs have been simplified by checking out the full repo.
This makes the jobs simpler, at the cost of slower performance.

KevinVlaanderen committed:
chore: combine multiple jobs with same triggers in one workflow

KevinVlaanderen committed:
chore: set since scope only once

KevinVlaanderen committed:
chore: cancel running workflows for same ref

KevinVlaanderen committed:
chore: combine linting and unit tests; fail workflow if checks fail

KevinVlaanderen committed:
chore: upload widgets as artifacts

KevinVlaanderen committed:
chore: update lerna to v5 and enable nx

KevinVlaanderen committed:
chore: (re)store nx cache in workflow

iobuhov commented:
I'm not reviewed this deeply, but I already have two concerns:
1. Why we removing `npm install`? As I see `npm install` was required to install `lerna` itself.
2. Nx and lerna 5 integration is still in experimental... don't you think it's  too early to add nx?

KevinVlaanderen committed:
chore: don't run script tests on node 14 anymore

KevinVlaanderen commented:
@iobuhov I'm not sure what you mean with question 1? `lerna bootstrap` does `npm install` under the hood. In the old case, we actually ran `npm install`, which would at the end trigger `lerna bootstrap`, which would then run `npm install` again. So that wasn't very efficient.
As for 2, yes, that's true. Everything up until 3 commits ago was irrespective of the lerna upgrade. Right now, I'm trying it out to see how it performs. We can discuss whether it makes sense to move.

iobuhov commented:
> I'm not sure what you mean with question 1? lerna bootstrap does npm install under the hood. In the old case, we actually ran npm install, which would at the end trigger lerna bootstrap, which would then run npm install again. So that wasn't very efficient.

I agree that this part was inefficient. My point was that we have to use `npm install` before `lerna bootstrap` to actually install `lerna` package itself. But now I see that you use `npx lerna bootstrap` which will install `lerna` and then run `lerna bootstrap` so nothing will break.

iobuhov commented on .github/workflows/Build.yml:
```diff
@@ -46,6 +46,14 @@ jobs:
+          path: node_modules/.cache/nx
+          key: cache-nx-${{ hashFiles('yarn.lock') }}-${{ matrix.target }}-${{ matrix.jobIndex }}-${{ github.sha }}
+          restore-keys: |
+            cache-nx-${{ hashFiles('yarn.lock') }}-${{ matrix.target }}-${{ matrix.jobIndex }}-
+            cache-nx-${{ hashFiles('yarn.lock') }}-
```
Why this two ends with `-`?

iobuhov replied:
And where this `yarn.lock` comes from?

KevinVlaanderen replied:
You're right, this was a poor copy-paste. I've adjusted it now.

iobuhov commented:
I have small request -- can we have `description` for jobs? This will simplify reading and understanding of what this job does.

iobuhov commented on .github/workflows/Build.yml:
```diff
@@ -1,61 +1,127 @@
+    steps:
+      - id: since
+        if: github.event_name == 'pull_request'
+        run: echo "::set-output name=since::-- --since origin/${{ github.base_ref }}"
```
I think meybe it's better to ermove `-- --since` from this output strgin?
I think this part should be part of command you run on step:
Now we have `npm run test ${{ needs.scope.outputs.since }}`
and it's not so clear, what is inside `needs.scope.outputs.since`
If we convert it to `npm run test -- --since ${{ needs.scope.outputs.since }}` it tells you that variable is expected to be a `--since` flag value.

BTW, it's just POV

KevinVlaanderen committed:
chore: fix file permissions

KevinVlaanderen committed:
chore: fix OoM error

KevinVlaanderen committed:
fix: update shelljs due to issue in mkdir

Shelljs 0.8.4 contains a bug, where it tried to create the target dir
non-recursively if the base dir exists.
However, if the target dir also already exists, this would fail.
This issue manifested in packages/shared/charts/src/rollup/index.js:81

KevinVlaanderen committed:
fix: disable plotly/ace caching for now; it seems to lead to concurrency issues

iobuhov commented on .github/workflows/AtlasReleaseThemeFiles.yml:
```diff
@@ -5,63 +5,56 @@ name: Create Release For Atlas Theme Folder
 # upgraded to Atlas 3, a migration doc included a link to a .zip file with theme files.
 
```
Same here, @jordanch , @pratikgoyal9 , can we add high level comment what this job does? Why we have to release Atlas Theme Files?

iobuhov commented on .github/workflows/Build.yml:
```diff
@@ -1,61 +1,125 @@
-          matrix:
-            os: [ ubuntu-latest, windows-latest ]
+  scope:
+    name: Set scope
```
```suggestion
    name: Set lerna scope with since flag
    # Only run build script in packages that have been changed compared to master
```

iobuhov commented on .github/workflows/Build.yml:
```diff
@@ -1,61 +1,125 @@
+      - name: Increase number of watchers
+        if: startsWith(matrix.os, 'ubuntu')
+        run: echo fs.inotify.max_user_watches=524288 | sudo tee -a /etc/sysctl.conf && sudo sysctl -p
+      - name: "Running build for development"
```
```suggestion
      - name: "Running pluggable-widgets-tools commands test"
```

iobuhov commented on .github/workflows/Build.yml:
```diff
@@ -1,61 +1,125 @@
+        with:
+          name: widgets
+          path: packages/pluggableWidgets/**/dist/*/*.mpk
+  script-tests:
+    name: "Script tests (Node ${{ matrix.node }}, ${{ matrix.os }})"
```
```suggestion
    name: "PWT commands test (Node ${{ matrix.node }}, ${{ matrix.os }})"
```

iobuhov replied:
```suggestion
  pluggable-widgets-tools-tests:
    name: "PWT tests (Node ${{ matrix.node }}, ${{ matrix.os }})"
```

iobuhov commented on .github/workflows/WebAutomatedTests.yml:
```diff
@@ -1,89 +1,88 @@
+          path: node_modules/.cache/nx
+          key: cache-nx-${{ hashFiles('package-lock.json') }}-${{ github.sha }}
+          restore-keys: |
+            cache-nx-${{ hashFiles('package-lock.json') }}-
```
```suggestion
            cache-nx-${{ hashFiles('package-lock.json') }}
```
